### PR TITLE
LG-3940: NumberInput onBlur runs with arrow buttons

### PR DESCRIPTION
**Pull request by stephl3**
## ✍️ Proposed changes

In the `NumberInput` component, the `onBlur` callback was not being invoked when the arrow buttons were clicked. The `onBlur` was being passed in with the rest of the props directly to the input, and this change will now pass it to both the input and the arrows.

🎟 _Jira ticket:_ [LG-3940](https://jira.mongodb.org/browse/LG-3940)
  _Linked to_: [BAAS-27278](https://jira.mongodb.org/browse/BAAS-27278) 

### For bug fixes, new features & breaking changes

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

## 🧪 How to test changes

I added storybook changes, so this can be tested by adding/removing focus on the arrow buttons and reviewing the console log entries.

https://github.com/mongodb/leafygreen-ui/assets/52300529/b3050fb6-565d-4c4d-bf4c-1781f969d999

### CODE REVIEW

File: packages/number-input/src/Input/Input.tsx [hunk 1]
```diff
@@ -162,23 +163,23 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
...
     };
 
-    const handleOnFocus = () => {
+    const handleFocusContainer = () => {
       isFocusedRef.current = true;
       handleSetErrorTransition();
     };
 
-    const handleOnBlur = () => {
+    const handleBlurContainer = () => {
```
Comment (stephl3) [inline packages/number-input/src/Input/Input.tsx:171]: made these more explicit to differentiate with `onBlur` prop passed into the api

File: .changeset/short-games-double.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@leafygreen-ui/number-input': patch
+---
+
+onBlur runs with arrow buttons
```
Comment (TheSonOfThomp) [inline .changeset/short-games-double.md:5]: Can we be more explicit with the change notes? What was the problem before? How does this impact consumers? 
Also, we've been starting to add the JIRA ticket into the change notes as well

File: packages/number-input/src/Input/Input.tsx (hunk 1 above)
Comment (TheSonOfThomp) [reply packages/number-input/src/Input/Input.tsx:171]: nitpick: should these be `handleContainerBlur/Focus` instead?

File: packages/number-input/src/Input/Input.tsx [hunk 3]
```diff
@@ -162,23 +163,23 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
       }
     };
 
-    const handleOnFocus = () => {
+    const handleFocusContainer = () => {
```
Comment (TheSonOfThomp) [inline packages/number-input/src/Input/Input.tsx:166]: Could we resolve this by listening to the focus/blur event that bubbles up from the input or button, instead of adding a listener to each element?

**Review by TheSonOfThomp — CHANGES_REQUESTED** [review]

File: packages/number-input/src/Input/Input.tsx (hunk 3 above)
Comment (stephl3) [reply packages/number-input/src/Input/Input.tsx:166]: thanks for this feedback, I like this approach better

File: .changeset/short-games-double.md (hunk 2 above)
Comment (stephl3) [reply .changeset/short-games-double.md:5]: updated!

File: packages/number-input/src/Input/Input.tsx [hunk 4]
```diff
@@ -162,23 +164,26 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
...
+    const handleFocus = () => {
       isFocusedRef.current = true;
       handleSetErrorTransition();
     };
 
-    const handleOnBlur = () => {
+    const handleBlur = (e: FocusEvent<HTMLDivElement>) => {
       isFocusedRef.current = false;
       handleRemoveErrorTransition();
+      if (onBlur) {
```
Comment (shaneeza) [inline packages/number-input/src/Input/Input.tsx:175]: I think we can shorten this to `onBlur?.(e);`

File: packages/number-input/src/Input/Input.tsx [hunk 5]
```diff
@@ -162,23 +164,26 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
...
       <div
         ref={containerRef}
-        onMouseEnter={() => handleSetErrorTransition()}
-        onMouseLeave={() => handleRemoveErrorTransition()}
-        onFocus={() => handleOnFocus()}
-        onBlur={() => handleOnBlur()}
+        onMouseEnter={handleSetErrorTransition}
+        onMouseLeave={handleRemoveErrorTransition}
+        onFocus={handleFocus}
+        onBlur={e => handleBlur(e)}
```
Comment (shaneeza) [inline packages/number-input/src/Input/Input.tsx:186]: I think we can get away with just `onBlur={handleBlur}`

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx [hunk 6]
```diff
@@ -243,6 +230,39 @@ describe('packages/number-input', () => {
...
+          label,
+          ...defaultProps,
+        });
+
+        const upArrow = container.querySelector(
+          'button[aria-label="decrement number"]',
+        );
+
+        userEvent.click(upArrow as HTMLButtonElement); // focus
+        expect(defaultProps.onBlur).toHaveBeenCalledTimes(1);
```
Comment (shaneeza) [inline packages/number-input/src/NumberInput/NumberInput.spec.tsx:258]: Why is `onBlur` triggered when clicking on the `upArrow`? Was the div in focus prior to the click?

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx [hunk 7]
```diff
@@ -243,6 +230,39 @@ describe('packages/number-input', () => {
...
+      });
+
+      test('callback triggers when focus leaves arrow buttons', () => {
+        const { container, numberInput } = renderNumberInput({
+          label,
+          ...defaultProps,
+        });
+
+        const upArrow = container.querySelector(
```
Comment (shaneeza) [inline packages/number-input/src/NumberInput/NumberInput.spec.tsx:253]: Maybe we should add testids to the arrows so this its easier for consumers to test?

Comment (stephl3) [reply packages/number-input/src/NumberInput/NumberInput.spec.tsx]: great, added those test ids and updated the relevant specs

**Review by shaneeza — APPROVED** [review]

File: packages/number-input/src/Input/Input.tsx [hunk 8]
```diff
@@ -162,23 +164,28 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
...
+    const handleFocus = () => {
       isFocusedRef.current = true;
       handleSetErrorTransition();
     };
 
-    const handleOnBlur = () => {
+    const handleBlur = (e: FocusEvent<HTMLDivElement>) => {
       isFocusedRef.current = false;
       handleRemoveErrorTransition();
+
```
[the remark this answers is not in the document]
Comment (stephl3) [reply packages/number-input/src/Input/Input.tsx]: added! 
`// If newly focused element is a child of the input container, we do not invoke onBlur`

File: packages/number-input/src/Arrows/Arrow.tsx [hunk 9]
```diff
@@ -34,6 +34,7 @@ export const Arrow = ({
       type="button"
       tabIndex={-1} // Mimicking native behavior; you cannot focus on an arrow.
       disabled={disabled}
+      data-testid={`lg-number_input-${direction}-button`}
```
Comment (TheSonOfThomp) [inline packages/number-input/src/Arrows/Arrow.tsx:37]: super-nit: `_button`

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx [hunk 10]
```diff
@@ -133,8 +135,10 @@ describe('packages/number-input', () => {
     });
 
     test('value change triggers onChange callback', () => {
+      const onChange = jest.fn();
```
Comment (TheSonOfThomp) [inline packages/number-input/src/NumberInput/NumberInput.spec.tsx:138]: 👍

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx [hunk 11]
```diff
@@ -202,47 +206,64 @@ describe('packages/number-input', () => {
...
+          ...defaultProps,
+        });
+
+        const upArrow = getByTestId(arrowTestId.up);
+
+        userEvent.click(upArrow as HTMLButtonElement); // focus
+        expect(upArrow).toHaveFocus();
+
+        userEvent.tab(); // blur
+        expect(onBlur).toHaveBeenCalledTimes(1);
```
Comment (TheSonOfThomp) [inline packages/number-input/src/NumberInput/NumberInput.spec.tsx:263]: 👍

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx [hunk 12]
```diff
@@ -202,47 +206,64 @@ describe('packages/number-input', () => {
...
+        });
+
+        userEvent.tab(); // focus
+        expect(numberInput).toHaveFocus();
+        userEvent.tab(); // blur
+
+        expect(onBlur).toHaveBeenCalledTimes(1);
+      });
+
+      test('callback triggers when focus leaves arrow buttons', () => {
```
Comment (TheSonOfThomp) [inline packages/number-input/src/NumberInput/NumberInput.spec.tsx:249]: We should test the case(s) where `onBlur` should _not_ be called

**Review by TheSonOfThomp — COMMENTED** [review]
Just have the one comment on testing negative cases. Otherwise, LGTM!

File: packages/number-input/src/NumberInput/NumberInput.spec.tsx (hunk 12 above)
Comment (stephl3) [reply packages/number-input/src/NumberInput/NumberInput.spec.tsx:249]: makes sense, added a test case

**Review by TheSonOfThomp — APPROVED** [review]
